client/engine/messageservice/test-messageservice.go

```diff
@@ -0,0 +1,75 @@
+package messageservice
+
+import (
+	"fmt"
+
+	"github.com/statechannels/go-nitro/protocols"
+	"github.com/statechannels/go-nitro/types"
+)
+
+// TestMessageService is an implementaion of the MessageService interface
+// for use in with-peers style test environments.
+//
+// It allows for individual nitro-clients / engines to:
+//  1. be instantiated together via test setup data
+//  2. "connect" with one another via gochans
+//  3. run independently in information-silo goroutines, while
+//     communicating on the simulated network
+type TestMessageService struct {
+	address types.Address
+
+	toPeers map[types.Address]chan<- protocols.Message
+	out     chan protocols.Message
+
+	in chan protocols.Message
+}
+
+func (t TestMessageService) Run() {
+	go t.routeOutgoing()
+}
+
+func (t TestMessageService) GetReceiveChan() chan protocols.Message {
+	return t.in
+}
+
+func (t TestMessageService) GetSendChan() chan<- protocols.Message {
+	return t.out
+}
+
+func (t TestMessageService) Send(message protocols.Message) {
+	t.out <- message
+}
+
+// Connect creates a gochan for message service t to communicate with
+// the given peer. This connection is one-way.
+func (t TestMessageService) Connect(peer TestMessageService) {
+	toPeer := make(chan protocols.Message)
+
+	t.toPeers[peer.address] = toPeer
+
+	go func() {
+		for msg := range toPeer {
+			peer.in <- msg
+		}
+	}()
```

Comment:
I'm curious about this bit. If `toPeer` was just made, won't it have 0 messages in it? Meaning this code is pointless?

Reply:
I looked into this some more and answered my own question. Recording my learnings here for posterity:

>  the `range` loop receives values from the channel repeatedly until it is closed.

Since we aren't closing the channel, this is essentially an infinite loop taking values from the channel (which is what I expected). 

Source https://go.dev/tour/concurrency/4

Reply:
These constructs (range over a channel, select statements) are really pleasant!